src/pykka/_proxy.py

```diff
@@ -221,29 +220,110 @@ def __getattr__(self, name):
 
-    def __setattr__(self, name, value):
+    def setattr(self, name):
         """
         Set a field on the actor.
 
         Blocks until the field is set to check if any exceptions was raised.
         """
+        attr_path = self._attr_path + (name,)
+        return lambda value: messages.ProxySetAttr(attr_path=attr_path, value=value)
+
+
+class ActorProxyBase:
+    @classmethod
+    def from_actor_ref(cls, actor_ref):
+        return cls(ActorProxyMessageBuilder(actor_ref))
+
+    def __init__(self, message_builder):
+        self._message_builder = message_builder
+        self.actor_ref = message_builder.actor_ref
+
+    def __eq__(self, other):
+        if not isinstance(other, ActorProxyBase):
+            return False
+        else:
+            return self._message_builder == other._message_builder
+
+    def __hash__(self):
+        return hash(self._message_builder)
+
+    def __dir__(self):
+        result = dir(self.__class__) + self._message_builder.dir()
+        return sorted(result)
+
+    def __repr__(self):
+        return f"<ActorProxyBase with {self._message_builder!r}>"
+
+
+class ActorProxy(ActorProxyBase):
+    def __getattr__(self, name):
+        res = self._message_builder.getattr(name)
+        if isinstance(res, CallableProxyFactory):
+            return res()
+        elif isinstance(res, ActorProxyMessageBuilder):
+            return ActorProxy(res)
+        else:
+            return self.actor_ref.ask(res, block=False)
+
+    def __setattr__(self, name, value):
         if name == "actor_ref" or name.startswith("_"):
             return super().__setattr__(name, value)
-        attr_path = self._attr_path + (name,)
-        message = messages.ProxySetAttr(attr_path=attr_path, value=value)
-        self.actor_ref.ask(message)
+        message_factory = self._message_builder.setattr(name)
+        self.actor_ref.ask(message_factory(value))
+
+
+class ExtendedActorProxy:
+    def __init__(self, actor_ref):
+        self._message_builder = ActorProxyMessageBuilder(actor_ref)
+        self.actor_ref = actor_ref
+
+    def delayed(self, delay):
+        return DelayedProxy(self._message_builder, delay=delay)
+
+
+class DelayedProxy(ActorProxyBase):
+    def __init__(self, message_builder, delay=0):
+        self._message_builder = message_builder
+        self._delay = delay
+        self.actor_ref = message_builder.actor_ref
+
+    def __getattr__(self, name):
+        res = self._message_builder.getattr(name)
+        if isinstance(res, CallableProxyFactory):
+            return res(delay=self._delay)
+        elif isinstance(res, ActorProxyMessageBuilder):
+            return DelayedProxy(res, delay=self._delay)
+        else:
+            self.actor_ref.ask(res, block=False, delay=self._delay)
+
+    def __setattr__(self, name, value):
+        if name == "actor_ref" or name.startswith("_"):
+            return super().__setattr__(name, value)
+        message_factory = self._message_builder.setattr(name)
```

Comment:
Reading this line I almost assumed it was changing an attr on the builder, but I assume it's creating a builder that has the job of setting an attr? Could we make this more clear with better naming or comments?